### PR TITLE
SQN-718-2

### PR DESCRIPTION
**Pull request by lamcc21**
- Update PurchaseOffer and ServiceAgreement handlers 
- Update parameters for OfferAccepted event 
- Update contract addresses
- Update contract sdk version

### CODE REVIEW

File: src/mappings/serviceAgreement.ts [hunk 1]
```diff
@@ -15,8 +15,10 @@ export async function handleServiceAgreementCreated(
   logger.info('handleClosedServiceAgreementCreated');
   assert(event.args, 'No event args');
 
+  const eventServiceAgreementId = event.args.serviceAgreementId.toString();
+
   const saContract = IServiceAgreement__factory.connect(
-    event.args.serviceAgreement,
+    eventServiceAgreementId,
     new FrontierEthProvider()
   );
```
Comment (lachlan1m) [inline src/mappings/serviceAgreement.ts:24]: This shouldn't be `event.args.serviceAgreementId.toHexString` right?

File: src/mappings/serviceAgreement.ts [hunk 2]
```diff
@@ -15,8 +15,10 @@ export async function handleServiceAgreementCreated(
   logger.info('handleClosedServiceAgreementCreated');
   assert(event.args, 'No event args');
 
+  const eventServiceAgreementId = event.args.serviceAgreementId.toString();
```
Comment (lamcc21) [inline src/mappings/serviceAgreement.ts]: Should this be event.args.serviceAgreementId.toHexString() in order to connect with factory?

File: src/mappings/purchaseOffer.ts [hunk 3]
```diff
@@ -59,9 +59,17 @@ export async function handlePurchaseOfferAccepted(
   assert(event.args, 'No event args');
 
   const eventOfferId = event.args.offerId.toString();
+  const eventAgreementId = event.args.agreementId.toString();
+
   const offer = await Offer.get(eventOfferId);
   assert(offer, `offer not found. offerID="${eventOfferId}"`);
 
+  const serviceAgreement = await ServiceAgreement.get(eventAgreementId);
```
Comment (mzxyz) [inline src/mappings/purchaseOffer.ts]: Better not check this.  Typically `OfferAcceptedEvent` should behind the `ServiceAgreementCreated` event, there still have possibility that when the assert execute but the new agreement is not saved yet even though they are in the same transaction.

File: src/mappings/serviceAgreement.ts (hunk 2 above)
Comment (mzxyz) [reply src/mappings/serviceAgreement.ts]: No serviceAgreement factory need anymore, I mention this before that we don't create agreement as new contract. You can access the agreement information with the following code:

```
const agreementRegistry = IServiceAgreementRegistry__factory.connect(
    eventServiceAgreementId,
    new FrontierEthProvider()
  );
  
const agreement = await agreementRegistry.closedServiceAgreements(agreementId)
const { period, lockedAmount } = agreement;
```

File: src/mappings/serviceAgreement.ts [hunk 4]
```diff
@@ -30,7 +32,7 @@ export async function handleServiceAgreementCreated(
   endTime.setSeconds(endTime.getSeconds() + period.toNumber());
 
   const sa = ServiceAgreement.create({
-    id: event.args.serviceAgreement,
+    id: eventServiceAgreementId,
```
Comment (mzxyz) [inline src/mappings/serviceAgreement.ts]: This is the updated agreement structure, add the `planId` and `planTemplateId` in the entity as well

```
struct ClosedServiceAgreementInfo {
    address consumer;
    address indexer;
    bytes32 deploymentId;
    uint256 lockedAmount;
    uint256 startDate;
    uint256 period;
    uint256 planId;
    uint256 planTemplateId;
}
```

Comment (lamcc21) [reply src/mappings/serviceAgreement.ts]: Thanks, updated

File: src/mappings/serviceAgreement.ts (hunk 2 above)
Comment (lamcc21) [reply src/mappings/serviceAgreement.ts]: updated

File: src/mappings/serviceAgreement.ts [hunk 5]
```diff
@@ -15,29 +15,32 @@ export async function handleServiceAgreementCreated(
...
+    eventServiceAgreementId.toString(),
     new FrontierEthProvider()
   );
 
-  const [period, value] = await Promise.all([
-    saContract.period(),
-    saContract.value(),
-  ]);
+  const agreement = await agreementRegistry.getClosedServiceAgreement(
+    eventServiceAgreementId
```
Comment (mzxyz) [inline src/mappings/serviceAgreement.ts:26]: `eventServiceAgreementId` should be `agreementRegistry` contract address

Comment (lachlan1m) [reply src/mappings/serviceAgreement.ts:26]: ` const eventServiceAgreementId = event.args.serviceAgreementId.toString();

  const agreementRegistry = IServiceAgreementRegistry__factory.connect(
    SA_REGISTRY_ADDRESS,
    new FrontierEthProvider()
  );

  const agreement = await agreementRegistry.getClosedServiceAgreement(
    eventServiceAgreementId
  );`

Comment (lamcc21) [reply src/mappings/serviceAgreement.ts:26]:
```
 const eventServiceAgreementId = event.args.serviceAgreementId.toString();

  const agreementRegistry = IServiceAgreementRegistry__factory.connect(
    SA_REGISTRY_ADDRESS,
    new FrontierEthProvider()
  );

  const agreement = await agreementRegistry.getClosedServiceAgreement(
    eventServiceAgreementId
  );

Comment (lamcc21) [reply src/mappings/serviceAgreement.ts:26]: You mean it should be like this right?

File: src/mappings/serviceAgreement.ts [hunk 6]
```diff
@@ -15,29 +15,32 @@ export async function handleServiceAgreementCreated(
   logger.info('handleClosedServiceAgreementCreated');
   assert(event.args, 'No event args');
 
-  const saContract = IServiceAgreement__factory.connect(
-    event.args.serviceAgreement,
+  const eventServiceAgreementId = event.args.serviceAgreementId.toString();
```
Comment (mzxyz) [inline src/mappings/serviceAgreement.ts]:
```suggestion
  const eventServiceAgreementId = event.args.serviceAgreementId;
```

**Comment by mzxyz** [comment]
Awesome work!

File: schema.graphql [hunk 7]
```diff
@@ -226,7 +227,7 @@ type AcceptedOffer @entity {
   id: ID! # offerID:indexer
   indexer: Indexer!
   offer: Offer!
-  serviceAgreement: ServiceAgreement
+  serviceAgreement: ServiceAgreement!
```
Comment (heyacherry) [inline schema.graphql:230]: Might hit error at this part, when the insert speed not sync before the event trigger. Can keep in mind for this and validate later. @HZHAIX @lachlan1m

**Pull request merged**